flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/AdaptedRestartPipelinedRegionStrategyNG.java

```diff
@@ -0,0 +1,295 @@
+						resetTasks(unmodifiedVertices, globalModVersion);
+
+						// re-schedule tasks
+						rescheduleTasks(unmodifiedVertices, globalModVersion);
+					} catch (GlobalModVersionMismatch e) {
+						// happens when a global recovery happens concurrently to the regional recovery
+						// just stop this local failover
+					} catch (Exception e) {
+						throw new CompletionException(e);
+					}
+				})
+			.whenComplete(
+				(Object ignored, Throwable t) -> {
+					// fail globally if any error happens
+					if (t != null) {
+						LOG.info("Unexpected error happens in region failover. Fail globally.", t);
+						failGlobal(t);
+					}
+				});
+	}
+
+	@VisibleForTesting
+	protected CompletableFuture<?> cancelTasks(final Set<ExecutionVertexID> vertices) {
+		final List<CompletableFuture<?>> cancelFutures = vertices.stream()
+			.map(this::cancelExecutionVertex)
+			.collect(Collectors.toList());
+
+		return FutureUtils.combineAll(cancelFutures);
+	}
+
+	private void resetTasks(final Set<ExecutionVertex> vertices, final long globalModVersion) throws Exception {
+		final Set<CoLocationGroup> colGroups = new HashSet<>();
+		final long restartTimestamp = System.currentTimeMillis();
+
+		for (ExecutionVertex ev : vertices) {
+			CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
+			if (cgroup != null && !colGroups.contains(cgroup)){
+				cgroup.resetConstraints();
+				colGroups.add(cgroup);
+			}
+
+			ev.resetForNewExecution(restartTimestamp, globalModVersion);
+		}
+
+		// if there is checkpointed state, reload it into the executions
+		if (executionGraph.getCheckpointCoordinator() != null) {
+			// abort pending checkpoints to
+			// i) enable new checkpoint triggering without waiting for last checkpoint expired.
+			// ii) ensure the EXACTLY_ONCE semantics if needed.
+			executionGraph.getCheckpointCoordinator().abortPendingCheckpoints(
+				new CheckpointException(CheckpointFailureReason.JOB_FAILOVER_REGION));
+
+			final Map<JobVertexID, ExecutionJobVertex> involvedExecutionJobVertices =
+				getInvolvedExecutionJobVertices(vertices);
+			executionGraph.getCheckpointCoordinator().restoreLatestCheckpointedState(
+				involvedExecutionJobVertices, false, true);
+		}
+	}
+
+	private void rescheduleTasks(final Set<ExecutionVertex> vertices, final long globalModVersion) throws Exception {
+		final CompletableFuture<Void> newSchedulingFuture;
+		switch (executionGraph.getScheduleMode()) {
+
+			case LAZY_FROM_SOURCES:
+				newSchedulingFuture = AdaptedSchedulingUtils.scheduleLazy(vertices, executionGraph);
+				break;
+
+			case EAGER:
+				newSchedulingFuture = AdaptedSchedulingUtils.scheduleEager(vertices, executionGraph);
+				break;
+
+			default:
+				throw new JobException("Schedule mode is invalid.");
+		}
+
+		// if no global failover is triggered in the scheduling process,
+		// register a failure handling callback to the scheduling
+		if (isLocalFailoverValid(globalModVersion)) {
```

Comment:
I think between `AdaptedSchedulingUtils.schedule[...]()` and this line, there cannot be modifications of the ExecutionGraph.

Reply:
This is to ensure that no global failure happens during the sync part of the schedule logic. Usually this should not happen. It's a insurance without much cost.

Reply:
Does it _usually not happen_, or is it forbidden to happen by contract?

I would rather add a check asserting that `rescheduleTasks()` is executed in the main thread. The code should be written so that it relies on well-defined contracts/invariants otherwise the next person who has to maintain the code might end up adding more paranoid `isLocalFailoverValid()` checks.

Here is more food for thought:
- Is it correct to add a null check after calling a method that is annotated with `@Nonnull`?
- Is it correct to add a null check after calling a method that returns `Optional`?

I think the same principle applies here.

Reply:
To clarify, this check is not for concurrent failovers. So main thread check does not help here. 

The main concern here is that some synchronize part of scheduling might cause task failure immediately. In this case, for example, before invoking `scheduleEager` the job is RUNNING and globalModVersion==1, after `scheduleEager ` returns, the job status is FAILING and globalModVersion==2. 

Although we can have contract that there should be no direct failing in the scheduling process, it can be hard to guarantee that. That's why I'd add a check here as an insurance.

Reply:
Ok, if it can happen, then we can leave it.